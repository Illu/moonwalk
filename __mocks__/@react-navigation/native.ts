import { Theme, DefaultTheme as defaultTheme } from "@react-navigation/native";

export const LightTheme = {
  ...defaultTheme,
  colors: {
    ...defaultTheme.colors,
    secondary: "#000",
    background: "#f1f2f7",
    text: "#000",
    placeholderText: "#7d7f86",
    inputBackground: "#dee1e7",
    secondaryText: "#666",
    accent: "#0a84ff",
    accentBackground: "rgba(10, 132, 255, 0.1)",
    uiAccent: "#c3c4c6",
  },
};

export const DefaultTheme = LightTheme;

export const DarkTheme: Theme = LightTheme;

export const useTheme = jest.fn().mockReturnValue({
  LightTheme,
});

export const useFocusEffect = jest.fn();
export const useNavigation = jest.fn().mockReturnValue({
  navigate: jest.fn(),
  replace: jest.fn(),
  reset: jest.fn(),
  setOptions: jest.fn(),
});
export const useNavigationState = jest.fn();
export const useRoute = jest.fn().mockReturnValue({
  params: {},
});
