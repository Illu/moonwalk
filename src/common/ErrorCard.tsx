import { useTheme } from "@react-navigation/native";
import React, { useState, useEffect } from "react";
import { Animated, Button, Easing } from "react-native";
import { TapGestureHandler } from "react-native-gesture-handler";
import styled from "styled-components/native";

import Icon from "../common/Icon";

const Wrapper = styled(Animated.View)`
  margin: 40px;
  border-radius: 10px;
  padding: 20px;
  background: ${({ theme }) => theme.colors.secondary};
`;

const Title = styled.Text`
  text-align: center;
  font-weight: bold;
  font-size: 20px;
  margin-left: 10px;
  color: ${({ theme }) => theme.colors.text};
`;

const Desc = styled.Text`
  text-align: center;
  margin: 10px;
  color: ${({ theme }) => theme.colors.text};
`;

const Detail = styled.Text`
  text-align: center;
  margin: 10px;
  color: ${({ theme }) => theme.colors.secondaryText};
  font-size: 11px;
  font-style: italic;
`;

const Row = styled.View`
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

interface Props {
  onRetry: () => any;
  message?: string;
  detail?: string | null;
}

const ErrorCard = ({ message, onRetry, detail }: Props) => {
  const { colors } = useTheme();
  const [appearAnim] = useState(new Animated.Value(0));

  useEffect(() => {
    Animated.timing(appearAnim, {
      toValue: 1,
      duration: 600,
      easing: Easing.out(Easing.quad),
      useNativeDriver: false
    }).start();
  }, [appearAnim]);

  return (
    <Wrapper
      style={{
        opacity: appearAnim,
        transform: [
          {
            translateY: appearAnim.interpolate({
              inputRange: [0, 1],
              outputRange: [80, 0],
            }),
          },
        ],
      }}
    >
      <Row>
        <Icon name="Alert" color={colors.text} />
        <Title>Error</Title>
      </Row>
      <Desc>{message || "Something went wrong"}</Desc>
      {detail && <Detail>{detail}</Detail>}
      <Button title="Try again" onPress={onRetry} />
    </Wrapper>
  );
};

export default ErrorCard;
