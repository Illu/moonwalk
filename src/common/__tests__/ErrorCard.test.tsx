import * as navigation from "@react-navigation/native";
import React from "react";

import TestRenderer from "../../helpers/testRenderer";
import { lightTheme } from "../../theme";
import ErrorCard from "../ErrorCard";

jest.mock("../../common/Icon", () => "Icon");
jest.useFakeTimers();
jest.spyOn(navigation, "useTheme");
navigation.useTheme.mockImplementation(() => lightTheme);

describe("ErrorCard", () => {
  it("renders correctly with a default message", () => {
    const props = {
      onRetry: jest.fn(),
    };
    const tree = TestRenderer(<ErrorCard {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders correctly with a custom message", () => {
    const props = {
      message: "$_MESSAGE_TEXT_$",
      onRetry: jest.fn(),
    };
    const tree = TestRenderer(<ErrorCard {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
