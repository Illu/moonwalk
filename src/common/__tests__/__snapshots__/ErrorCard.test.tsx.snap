// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorCard renders correctly with a custom message 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "borderRadius": 10,
      "marginBottom": 40,
      "marginLeft": 40,
      "marginRight": 40,
      "marginTop": 40,
      "opacity": 0,
      "paddingBottom": 20,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 20,
      "transform": Array [
        Object {
          "translateY": 80,
        },
      ],
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <Icon
      color="#000"
      name="Alert"
    />
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginLeft": 10,
            "textAlign": "center",
          },
        ]
      }
    >
      Error
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
          "marginBottom": 10,
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 10,
          "textAlign": "center",
        },
      ]
    }
  >
    $_MESSAGE_TEXT_$
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Try again
      </Text>
    </View>
  </View>
</View>
`;

exports[`ErrorCard renders correctly with a default message 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "borderRadius": 10,
      "marginBottom": 40,
      "marginLeft": 40,
      "marginRight": 40,
      "marginTop": 40,
      "opacity": 0,
      "paddingBottom": 20,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 20,
      "transform": Array [
        Object {
          "translateY": 80,
        },
      ],
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <Icon
      color="#000"
      name="Alert"
    />
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginLeft": 10,
            "textAlign": "center",
          },
        ]
      }
    >
      Error
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
          "marginBottom": 10,
          "marginLeft": 10,
          "marginRight": 10,
          "marginTop": 10,
          "textAlign": "center",
        },
      ]
    }
  >
    Something went wrong
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Try again
      </Text>
    </View>
  </View>
</View>
`;
